WebDevStudios/nextjs-wordpress-starter
# Feature/45 form handling

Part of https://github.com/WebDevStudios/nextjs-wordpress-starter/issues/45

### Link

Link to the branch preview generated by Vercel.

### Description

This PR adds a few major features in regards to handling Forms with and without GravityForms. Due to the sheer size of this task, this PR lays a lot of the ground work and is not feature complete. 

At this time, I've only mapped the `Text` input and will add more in separate PRs as more of the structure is in place.

**Note 1: Forms are not yet submittable but the data handling is in place.**

**Note 2: Fields that do not yet have a matching GravityForm Field component will automatically display with an "Unsupported Field" message in the content.**

- Setup a `<Form />` component that uses `Formik` and `Yup` for field validation, error, and state handling.
- Setup a `<GravityForm />` component that uses `<Form />` and routes GravityForm data.
- Setup an input system for form Input components. See `atoms/Inputs/Text`, ex `<Inputs.Text />`.
- Setup an input system for GravityForm field components. See `GravityForms/Fields`, ex `<GfFields.Text />`.
- Setup <Fields /> mapping component in `GravityForm/Fields/Fields.js`. This component is important for matching up GravityForm field types with components.
- Setup `<BlockGravityForm />` with steam-lined prop passing to prevent PropType declaration duplication. This component is for handling Block data from created by GravityForms and routes it to the GravityForm component.
- Setup dynamic field validation for Yup. Due to challenges architectural decisions of Formik it's really difficult to setup inline field validation. Instead this process occurs during the field mapping process since this component does not rerender frequently. Note: This is handled in a basic way and will require refactoring later. I left a TODO.

### Screenshot
This example shows a few variations of the "Text" field from GravityForms with various settings.
<img width="589" alt="Screen Shot 2020-12-31 at 6 34 45 PM" src="https://user-images.githubusercontent.com/25035900/103430932-e7bc7f80-4b96-11eb-9b86-4d41c951e86c.png">

### Verification
To fully test this feature you'll need to use a test branch since blocks are not implemented yet. 

1. Make sure you have the latest local pulled. This is important because you'll need the correct version of GravityForms GraphQL plugin.
2. Checkout `test/45-form-handling` and run dev.
3. Visit `http://localhost:3000/blog/form-input-test`
4. Test out the form inputs

## Form Testing
- Try adding or removing settings from the Input Test Form in Gravity Forms. Note: Not all settings work but some of the basics are wired up.
- Try modifying settings in the Text fields of the form. Change field requirements, add classes etc...
  - Some features in this area will be added later see unsupported field features

## Unsupported Field Features 
Some features are not completely supported at this time because they require more work or do not have data from GraphQL. See the following notes.

1. **Field Label** Visibility > Not available via GraphiQL at this time.
2. **Admin Label** This does not need to be included since this feature is mainly for dashboard purposes. We should probably exclude it from the query.
3. **AdminOnly** I'm not sure what this query param does at the time as changes to the field did not indicate a change.
4. **allowsPrepopulate** matches up with the Advanced "Allow field to be populated dynamically" field option. This should be built out later.
5. **conditionalLogic** Conditional display logic that works using Advanced > "Enable Conditional Logic" will need to be explored later.
6. **Field Visibility** Advanced > Visibility adds a classname to the component wrapper at this time. We may need to add additional features for loggedin users.
7. **errorMessage**: This will need circled back to as it may not be a necessary field since validation will occur with custom messages.
10. **noDuplicates** This will need circled back to later.
